tikv-client/src/main/java/com/pingcap/tikv/TiBatchWrite2PCClient.java

```diff
@@ -0,0 +1,555 @@
+          size++;
+        } else {
+          break;
+        }
+      }
+      doCommitSecondaryKeys(backOffer, keyBytes, size, commitTs);
+    }
+  }
+
+  private void doCommitSecondaryKeys(
+      BackOffer backOffer, ByteString[] keys, int size, long commitTs)
+      throws TiBatchWriteException {
+    if (keys == null || keys.length == 0 || size <= 0) {
+      return;
+    }
+
+    // groups keys by region
+    GroupKeyResult groupResult = this.groupKeysByRegion(backOffer, keys, size);
+    if (groupResult.hasError()) {
+      LOG.warn("commit secondary key error: " + groupResult.getErrorMsg());
+      return;
+    }
+
+    boolean sizeKeyValue = false;
+    List<BatchKeys> batchKeyList = new LinkedList<>();
+    Map<Long, List<ByteString>> groupKeyMap = groupResult.getGroupsResult();
+    for (Long regionId : groupKeyMap.keySet()) {
+      this.appendBatchBySize(
+          batchKeyList,
+          regionId,
+          groupKeyMap.get(regionId),
+          sizeKeyValue,
+          TXN_COMMIT_BATCH_SIZE,
+          null);
+    }
+
+    // For prewrite, stop sending other requests after receiving first error.
+    for (BatchKeys batchKeys : batchKeyList) {
+      doCommitSecondaryKeySingleBatch(backOffer, batchKeys, commitTs);
+    }
+    return;
+  }
+
+  private void doCommitSecondaryKeySingleBatch(
+      BackOffer backOffer, BatchKeys batchKeys, long commitTs) throws TiBatchWriteException {
+    List<ByteString> keysCommit = batchKeys.getKeys();
+    ByteString[] keys = new ByteString[keysCommit.size()];
+    keysCommit.toArray(keys);
+    // send rpc request to tikv server
+    long regionId = batchKeys.getRegioId();
+    ClientRPCResult commitResult =
+        this.kvClient.commit(backOffer, keys, this.startTs, commitTs, regionId);
+    if (!commitResult.isSuccess()) {
+      String error =
+          String.format("Txn commit secondary key error, regionId=%s", batchKeys.getRegioId());
+      LOG.warn(error);
+      throw new TiBatchWriteException("commit secondary key error: " + commitResult.getError());
+    }
+
+    // return success
+    LOG.info("commit " + batchKeys.getKeys().size() + "rows successfully");
+    return;
+  }
+
+  private GroupKeyResult groupKeysByRegion(BackOffer backOffer, ByteString[] keys, int size) {
+    Map<Long, List<ByteString>> groups = new HashMap<>();
+    long first = 0;
+    int index = 0;
+    String error = null;
+    try {
+      for (; index < size; index++) {
+        ByteString key = keys[index];
+        TiRegion tiRegion = this.regionManager.getRegionByKey(key);
+        if (tiRegion != null) {
+          Long regionId = tiRegion.getId();
+          if (index == 0) {
+            first = regionId;
+          }
+          groups.computeIfAbsent(regionId, e -> new LinkedList<>()).add(key);
+        }
+      }
+    } catch (Exception e) {
+      error = String.format("Txn groupKeysByRegion error, %s", e.getMessage());
+    }
+    GroupKeyResult result = new GroupKeyResult();
```

Comment:
it is a bit weird for GroupKeyResult to contain error message to indicate whether an exception is met. We might simply throw an Exception in this case to avoid checking if errorMsg is empty.

Reply:
fixed